docs/internal/interop.md

```diff
@@ -0,0 +1,351 @@
+generation) if the types involved are parametrised. All template arguments are
+taken into consideration and a specialisation is defined and generated in a
+wrapper layer.
+
+## Child process execution environment
+
+Our initial implementation is to separate the sandbox execution by creating a
+child process. This is not strictly speaking a sandbox, but gives us an easy
+way to test the rest of the framework without much additional work. Our aim is
+to use WebAssembly as our default sandbox environment, but that will take time
+and will be done later.
+
+The sandbox creation forks the process with a special sandbox library that will
+forward all calls to the parent process, which will control where the memory
+actually lives (see below). A local heap is created per sandbox for the child
+to use in its allocations.
+
+Calls will be forwarded across the parent/child barrier via the dispatcher, by
+means of a function pointer table, where the position is the index in the code
+generation stage above, and the buffer is the contents of the arguments and a
+place for the return value.
+
+Argument decomposition will be done by the sandbox code that will call the actual
+function with the actual arguments and, if there is a return value, set it on
+the right place of the buffer and return.
+
+### Memory Management
+
+To avoid the sandbox from requesting memory in the wrong places (accidentally
+or maliciously), we create a two-tier allocator. The actual allocator in the
+Verona side, and a dummy allocator on the sandbox side that redirects all its
+calls to the Verona allocator. This is part of the sandbox library's
+functionality, which is passed along on the creation of the sandbox.
+
+The Verona allocator creates a heap for the sandbox and that's the memory that
+the sandbox code can use for everything. The sandbox code can also access that
+memory and it uses to keep the arguments and return values, so that the sandbox
+can see them.
+
+Because that heap is in a Verona sandbox region, it is safe to assume there are
+no race conditions introduced by the compiler onto the external library, due to
+the use of cowns to execute the foreign code.
+
+## Step-by-step example
+
+This is the description of a mock implementation of the child process sandbox
+running the C++ snippet in the beginning of this document. We'll go line by line
+and describe what the compiler and the generated code will do for each of them.
+
+### Compile time
+
+```c
+type StdVec = "c++::vector";
+```
+
+* Call the C++ driver (clang wrapper) to parse the include file (and all its
+  includes) and keep as a C++ AST.
+* Create a Verona class called `StdVec` with C++ built-in types defined as Verona
+  types (ex. `int` as `S32`, `size_t` as `U64`).
+* Expose an interface for the parser to query symbols (types, objects and
+  functions).
+
+```c
+reg = new StdVec;
+```
+
+* The parser recognises the region as a C++ sandbox.
```

Comment:
The parser recognises the type as coming from a specific module.  It queries that module for types within that module as they are used.